Implement session restore

victorporof opened the pull request:
This is a big one. I'd like two sets of eyes on this. @jsantell @joewalker please take a look.

Known issue: after a session restore, typing in the location bar results in 500 internal server errors from the UAS when trying to fetch autocompletions.

joewalker commented:
Wait, I was only r+ing one changeset not the whole thing!?

victorporof commented:
@joewalker TOO LATE I'M SHIPPING!

jsantell commented on app/ui/browser-modern/model/profile.js:
```diff
@@ -12,6 +12,10 @@ specific language governing permissions and limitations under the License.
-export default Immutable.Record({
+import { register } from '../util/record-constructors';
+
+const VERSION = 1;
```
Do we have to increment this (and other model's version numbers) everytime any change is made to the state? Should have clear documentation on what the rules are for this so we don't run into a scenario where we're checking for "bad" version numbers of models

victorporof replied:
Should increment this every time we care about migrating the records (and it doesn't work automatically). That is to say: whenever we're changing things substantially (not just adding them) and care enough about not losing the state.

victorporof commented on app/ui/browser-modern/sagas/session-sagas.js:
```diff
@@ -0,0 +1,121 @@
+}
+
+function* getSessionKey({ key }) {
+  ipcRenderer.send('session:getkey', key);
```
True, this isn't used anywhere. I'll remove it, though the main process does indeed handle the `session:getkey` event, not sure what you mean by "it doesn't return it".

The `session:setkey` and `session:delkey` events are indeed handled.

victorporof replied:
Fixed.

ncalexan commented on app/main/menu/submenu-items.js:
```diff
@@ -140,6 +140,9 @@ items.reloadApp = {
   click(item, focusedWindow) {
     if (focusedWindow) {
       focusedWindow.reload();
+      focusedWindow.webContents.once('did-finish-load', () => {
```
Is this one right?  I don't expect to get a new tab when I reload.  Also, does "Reload App" reload _all_ open `BrowserWindows`?  If not, why not?  Should we re-label this to 'Reload Window' or similar?

victorporof commented on app/ui/browser-modern/sagas/external-sagas.js:
```diff
@@ -39,9 +42,20 @@ function* maximizeWindow() {
+  // Wait for any additional operations to finish before reloading the window.
+  // Add your own here if needed.
+  yield take(EffectTypes.NOTIFY_BROWSER_WINDOW_APP_STATE_SAVED);
+  ipcRenderer.send('guest-window-id', yield select(getWindowId));
```
`yield` always halts execution, but doesn't imply asynchronicity.

victorporof replied:
See https://yelouafi.github.io/redux-saga/docs/api/index.html#selectselector-args

victorporof commented on app/main/session-io.js:
```diff
@@ -0,0 +1,73 @@
+}
+
+ipcMain.on('session:getkey', (event, key) => {
+  getSessionKey(key);
```
Whoops, it should indeed, however, this isn't used anywhere at the moment in the main process, so I'll remove it.

victorporof replied:
Fixed.

ncalexan commented on app/ui/browser-modern/actions/page-effects.js:
```diff
@@ -17,21 +17,22 @@ import * as EffectTypes from '../constants/effect-types';
 import * as PagesSelectors from '../selectors/pages';
 import PageUIState from '../model/page-ui-state';
 
-export function createPageSession(location, options) {
```
nit: in commit message, 'malleable' has two 'l' characters.

victorporof replied:
Fixed.

ncalexan commented on app/ui/browser-modern/actions/page-effects.js:
```diff
@@ -17,21 +17,22 @@ import * as EffectTypes from '../constants/effect-types';
 }
 
-export function destroyPageSession(pageId) {
+export function destroyPageSession(pageId, { withUI } = {}) {
```
Is it worth hinting that `withUI` is a boolean here by defaulting it to `false`?

victorporof replied:
We're not defaulting `withUI` here to anything explicitly, but simply destructuring a parameter. The defaulting is done in the sagas (to `true`), not in the effect creators here. I understand the confusion though, so I'll reverse this: default it here and don't do anything special in the sagas.

victorporof replied:
Fixed.

ncalexan commented on app/ui/browser-modern/sagas/page-sagas.js:
```diff
@@ -49,18 +50,25 @@ export default function*() {
 }
 
-function* createPageSession({ id, location, options }) {
+function* createPageSession({ id, location, options, withUI = true }) {
```
Hmm.  Looking ahead, I expect the "bulk" operations to diverge pretty rapidly from the "respond to UI action" operations.  Would it be wiser to factor this into a "replacePageSessions" bulk action, and especially to add a "destroyPageSessions" action, for when we implement "Close All Tabs" / "Close Other Tabs", etc?

Also, it looks like you're starting a new UAS session for each page you restore.  Are you sure that's what you want?  I think we should at least tell the UAS this is a session restore; seems like the kind of data we might want to use and/or surface to say, "Hey, this tab has been restored 7 times, should I snooze it?"  We might also want to give you a bulk UAS call for this, so you don't spawn 1000 UAS POSTs to session restore.

None of these necessarily block, but they're worth thinking about.

victorporof replied:
Indeed, a bulk action might be a better idea in this particular case. For `close all tabs` and `close other tabs`, composing multiple `destroyPageSession` actions together is straightforward, no need for any custom shenanigans.

victorporof replied:
Regarding starting a new UAS session, I assume a new "page session". This is correct, and something I need to look into. It's probably not what I want. Would it be better to have a `restore session` API endpoint on the UAS? A bulk call is also going to help a lot with performance. Let's talk about this.

Furthermore, as mentioned in https://github.com/mozilla/tofino/pull/1415#issue-183066175 and https://github.com/mozilla/tofino/pull/1415#issuecomment-253852543 there is a problem currently with fetching autocompletions after a session restore, and I assume this is a generalized problem causing other issues due to creating page sessions when they do actually exist already (we're not closing the UAS when reloading the browser for example, and the page session ids are reused). I'll investigate.

victorporof replied:
Added bulk actions for creating / destroying standalone sessions without any UI. The usage of the UAS in these sagas is expected to change to avoid spamming posts, and highlight the session restore aspect more appropriately.

ncalexan commented on app/main/session-io.js:
```diff
@@ -0,0 +1,73 @@
+const SESSION_FILE = path.join(PROFILE, 'session.json');
+logger.info(colors.blue('Using'), colors.green(SESSION_FILE));
+
+// Avoid synchronous (potentially frequent) read/writes to/from the filesystem
```
Should this be de-bounced as well, so that we're not constantly writing when doing bulk operations?

victorporof replied:
I'm not sure. It should definitely not be debouned though, since debouncing can potentially lead to indefinite wait times; I think you mean "throttle" instead. The frontend already throttles sending the app state to the main process for storage, so constantly writing to disk isn't a problem right now. Not sure whether or not it's useful to throttle on both ends at this point.

ncalexan commented on app/main/session-io.js:
```diff
@@ -0,0 +1,73 @@
+  setSessionKey(key, value);
+});
+
+ipcMain.on('session:delkey', (event, key) => {
```
I haven't seen the use cases for this, but: if you expect multiple processes to share a single key-space, this invites cross-process race conditions.  If you expect each process to have its own key-space, perhaps you should enforce this, using something like the window ID you added in an earlier commit?

ncalexan commented on app/ui/browser-modern/util/record-constructors.js:
```diff
@@ -0,0 +1,22 @@
+  return record;
+}
+
+export function getAllRegisteredRecords() {
```
Should this return a clone or an immutable set?

victorporof replied:
100% meh. I'll convert it to an array, but I expect us to be smart enough not to modify the array returned by `getAllRegisteredRecords`. That being said, future versions of ourselves might not be careful, so I'll clone it.

ncalexan commented on package.json:
```diff
@@ -63,6 +63,8 @@
     "redux-saga": "0.12.0",
     "source-map-support": "ncalexan/node-source-map-support#fileUrls-plus",
     "tmp": "0.0.29",
+    "transit-immutable-js": "0.6.0",
```
I expect you want `transit` to rehydrate Immutable.js Records, but:

- is Transit's output itself stable across the life of the Tofino project/product?
- weigh heavily that Transit's output is not "hand hackable".  It's text, yes -- JSON, IIUC -- but it's not human readable or human editable.  At least, that's what I remember the last time I looked at some.

Is it worth being simpler here and just writing and reading JSON that _isn't_ Transit mangled?

victorporof replied:
We can definitely naively serialize the entire app state as json. After all, immutable objects are convertible to plain js objects. And indeed the json output from `transit` is very difficult to read by humans.

That being said, not doing so means writing custom deserializers from json to records, since we can't just easily use built-in immutable methods when converting back. To elaborate, creating an immutable.js objects from json isn't guaranteed to be the exact same as the original serialized immutable.js object. This is because immutable.js has higher level data structure types (such as sets and lists which convert to plain arrays in json), so information about the data types is lost when converting to json.

In short, if we don't use `transit` for serialization, we'll have to implement the same functionality ourselves if we want to restore the entire app state. If not, we abandon the entire idea and don't dump the app state to disk to begin with, but some 'view' of it.

ncalexan commented on app/ui/browser-modern/util/session-util.js:
```diff
@@ -0,0 +1,22 @@
+import { getAllRegisteredRecords } from './record-constructors';
+
+export function serializeAppState(state) {
+  // Save the entire app state for now. If for some reason we need to
```
Consider making this deterministic, by (say) sorting the record types, sorting the records themselves, and using a deterministic JSON outputter.  Testing might appreciate it.

Or is Transit deterministic?

victorporof commented:
Will address comments and land. Need to investigate the UAS errors when fetching autocompletions as well.

ncalexan commented on app/ui/browser-modern/sagas/session-sagas.js:
```diff
@@ -0,0 +1,91 @@
+function* watchAppStateChanges(fn) {
+  // It's safe to watch all actions here since we're throttling and not just
+  // debouncing (potentially indefinitely if actions are dispatched constantly
+  // with a lower frequency than the debounce delay). No reason to watch effects,
```
Effect vs. action -- this is actually pretty nice!

ncalexan commented on app/ui/browser-modern/sagas/session-sagas.js:
```diff
@@ -0,0 +1,91 @@
+  // Invoked periodically when changes are made to the app state.
+  const store = yield select(serializeAppState);
+  const windowId = yield select(getWindowId);
+  yield put(SessionEffects.setSessionKey(['browserWindows', windowId], store));
```
Ah, I see, you are expecting the *consumer* to partition the key-space.  And you're not doing it at the top level.  I imagine your "object-path" library avoids the obvious race between two setters, so this probably works.  In any case, considering stating the contract for consumers in the storage module itself, and consider moving the `windowId` to the top-level in some way, so there can be no conflict across windows.

victorporof replied:
I'll add comments where the API is defined. Currently the `windowId` is top level, part of the app state itself.

victorporof replied:
Fixed.

ncalexan commented on app/ui/browser-modern/sagas/session-sagas.js:
```diff
@@ -0,0 +1,91 @@
+import * as EffectTypes from '../constants/effect-types';
+import * as SessionEffects from '../actions/session-effects';
+
+const THROTTLE_DURATION = 1000; // ms
```
nit: comment on what this really means.  Is it that we'll never write more than once every second?

victorporof replied:
Yes it is. I'll add a comment.

victorporof replied:
Fixed.

ncalexan commented on app/ui/browser-modern/sagas/session-sagas.js:
```diff
@@ -0,0 +1,91 @@
+    takeLatest(...wrapped(EffectTypes.DEL_SESSION_KEY, delSessionKey)),
+  ];
+
+  // Make sure we're cancelling the `watchAppStateChanges` task when either
```
I don't understand the mechanics of sagas well enough to understand how this works.  @jsantell (or @joewalker), please ensure that you do!

ncalexan commented on app/ui/browser-modern/sagas/external-sagas.js:
```diff
@@ -40,10 +41,16 @@ function* maximizeWindow() {
 function* closeWindow() {
+  // Wait for any additional operations to finish before closing the window.
+  // Add your own here if needed.
+  yield take(EffectTypes.NOTIFY_BROWSER_WINDOW_APP_STATE_DESTROYED);
```
Is this a situation where you want 'before-close-browser-window' and then 'close-browser-window'?  Can a window _deny_ it's closing with this change set?  (I think it can't.)  Can a window do some potentially expensive IO operation to prepare it's saved state before closing?  (I think it can.)

victorporof replied:
Yes, this is a situation where some stuff needs to happen before a browser window closes. The logic for what happens is defined somewhere else, since I wouldn't like polluting sagas with unrelated application logic. The `take` operations make sure other things happen before continuation is allowed.

To answer your questions: 
1. A window can indeed deny its closing now with these changes, since requests for closing windows need to go through the renderer process and through this saga first. As the comments imply, the logic for the expected state / actions that need to happen beforehand can be added here.

2. A window can decide to do whatever it wants before its closing. In fact, this is already the case: when a window is closed manually, it is removed from the session storage. As another example, in the next saga in this file, when a window is reloaded, it requests that its app state is restored from the session.

ncalexan commented on app/main/menu/submenu-items.js:
```diff
@@ -140,8 +141,11 @@ items.reloadApp = {
       focusedWindow.webContents.send('reload-window');
-      focusedWindow.webContents.once('did-finish-load', () => {
-        focusedWindow.webContents.send('new-tab');
+
```
This makes more sense.  (Ignore my earlier comment.)

ncalexan commented on app/ui/browser-modern/reducers/all.js:
```diff
@@ -0,0 +1,27 @@
+specific language governing permissions and limitations under the License.
+*/
+
+import * as ActionTypes from '../constants/action-types';
```
Is it worth providing some hints as to what these reducers allow (do anything to the whole state!) and why it's necessary?

victorporof replied:
Doubt it.

ncalexan commented on app/main/menu/submenu-items.js:
```diff
@@ -140,8 +141,11 @@ items.reloadApp = {
-      focusedWindow.webContents.once('did-finish-load', () => {
-        focusedWindow.webContents.send('new-tab');
+
+      electron.ipcMain.once('guest-window-id', (event, id) => {
```
Is this right?  If `guest-window-id` is coming from Electron (I think it is, 'cuz I don't see it in Tofino), are we ceding ID generation to Electron?  Do we expect Electron's ID generation to have good properties?  (I.e., why isn't this just 1, 2, 3... each time we restart?)

victorporof replied:
No, we don't ceding ID generation to Electron. The `guest-window-id` message is coming from the renderer process: each one generates its own window id. Every time when loading browser windows (when creating them or reloading them), a new id is generated. Session restore on the browser window after a reload is taken care of here.

The fact that we have ipc messaging here is a manifestation of the general problem described in https://github.com/mozilla/tofino/issues/1354 and https://github.com/mozilla/tofino/issues/1414.

victorporof replied:
I'be fixed this by having the main process decide on a window id, instead of the renderer process. This implies less ad-hoc communication between them.

ncalexan commented on app/ui/browser-modern/model/active-element.js:
```diff
@@ -14,10 +14,12 @@ import Immutable from 'immutable';
 
+const VERSION = 1;
+
 const ActiveElement = register(Immutable.Record({
```
I feel like you could work the versioning into `register`, and make this more uniform.

victorporof replied:
Not sure how. The record naming is part of the record constructor, not the `register` function.

ncalexan commented on app/ui/browser-modern/util/session-migrate.js:
```diff
@@ -0,0 +1,54 @@
+// taking the data to migrate as a basic Immutable.Map, and having to return it
+// in the new format.
+const MIGRATIONS = {
+  // e.g. migrate MyRecord from v1 to v2:
```
Perhaps worth including how to sequence migrations.  That is, if you get `MyRecord_v1` but you expect v3, you have to update `MyRecord_v1` and make `MyRecord_v2` call `MyRecord_v1`, etc.

ncalexan commented:
Any details here?  We do move the UAS forward, so we might have broken
something on you...

On Fri, Oct 14, 2016 at 9:32 AM, Victor Porof <notifications@github.com>
wrote:

> Will address comments and land. Need to investigate the UAS errors when
> fetching autocompletions as well.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/tofino/pull/1415#issuecomment-253852543>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABC93gmtocMtipRuNYR59gh_nZgkz4_Qks5qz66ggaJpZM4KXFAB>
> .
>

victorporof commented on app/ui/browser-modern/actions/session-effects.js:
```diff
@@ -0,0 +1,60 @@
+  };
+}
+
+export function delSessionKey(key) {
```
Fixed.

victorporof commented:
Addressed all comments.

victorporof commented:
I've made some relatively substantial changes after addressing all review comments. @ncalexan @jsantell @joewalker please take another look if you have the time.

I've arranged the commits to tell a story, following them can aid the review.

victorporof commented:
@ncalexan Regarding the UAS errors, I'm still investigating.

victorporof commented:
@ncalexan The UAS errors when trying to fetch autocompletions are unrelated. I've filed https://github.com/mozilla/tofino/issues/1425

victorporof commented:
Since the UAS errors are unrelated, I'm landing this.
